core/templates/dev/head/pages/exploration-editor-page/improvements-tab/services/improvement-modal.service.ts

```diff
@@ -48,6 +48,177 @@ angular.module('oppia').factory('ImprovementModalService', [
+                    // current learner action differ, they are part of
+                    // different contexts. So, we identify whether the size
+                    // of the block has crossed the threshold number of actions
+                    // per block. If it has, we don't add more learner actions
+                    // to the block.
+                    if (indexOfFirstDisplayedAction - i >= (
+                      MAX_UNRELATED_ACTIONS_PER_BLOCK)) {
+                      indexOfFirstDisplayedAction = i + 1;
+                      break;
+                    }
+                  }
+                  previousStateName = currentStateName;
+                }
+                // This is the case when all learner actions have been iterated
+                // over, and no more expansion is possible.
+                if (i === -1) {
+                  indexOfFirstDisplayedAction = 0;
+                }
+              };
+
+              expandActions();
+
+              // Index to know where to start highlighting actions.
+              let indexOfActionInFirstDisplayedBlock =
+                indexOfFirstDisplayedAction;
+
+              $scope.issueIsMultipleIncorrectSubmissions = false;
+              if (playthrough.issueType === (
+                ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS)) {
+                $scope.issueIsMultipleIncorrectSubmissions = true;
+              }
+
+              /**
+               * Renders a table for MultipleIncorrectSubmissions issue for a
+               * more friendly UI.
+               */
+              $scope.renderIssueTable = function() {
+                var lars = LearnerActionRenderService;
+                var tableHtml =
+                  lars.renderFinalDisplayBlockForMISIssueHTML(
+                    playthrough.actions.slice(
+                      indexOfActionInFirstDisplayedBlock),
+                    indexOfActionInFirstDisplayedBlock + 1);
+                return tableHtml;
+              };
+
+              /**
+               * Returns the list of learner actions to currently be
+               * displayed in the playthrough modal.
+               * @returns {LearnerAction[]}
+               */
+              $scope.getActionsToRender = function() {
+                if ($scope.issueIsMultipleIncorrectSubmissions) {
+                  // This corresponds to the case where we display the table
+                  // for the issue.
+                  return playthrough.actions.slice(
+                    indexOfFirstDisplayedAction,
+                      indexOfActionInFirstDisplayedBlock);
+                }
+                return playthrough.actions.slice(indexOfFirstDisplayedAction);
+              };
+
+              /**
+               * Returns the index of the learner action.
+               * @param {LearnerAction} learnerAction.
+               * @returns {int}
+               */
+              $scope.getLearnerActionIndex = function(learnerAction) {
```

Comment:
Try not to have functions like this. Instead, work with actionIndexes throughout the code, rather than actions which are big and bulky. Translate actionIndexes to actions only at the last minute or on-demand.

Reply:
Done.